ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/LagReportingAgent.java

```diff
@@ -0,0 +1,380 @@
+        final Integer partition = partitionEntry.getKey();
+        final LagInfoEntity lagInfo = partitionEntry.getValue();
+        partitionsBuilder.put(partition, lagInfo);
+      }
+
+      hostMapBuilder.put(lagInfoKey, partitionsBuilder.build());
+    }
+
+    Map<LagInfoKey, Map<Integer, LagInfoEntity>> hostMap = hostMapBuilder.build();
+    HostLagInfo hostLagInfo = new HostLagInfo(hostMap, updateTimeMs);
+    receivedLagInfo.compute(hostInfo, (hi, previousHostLagInfo) ->
+        previousHostLagInfo != null && previousHostLagInfo.getUpdateTimeMs() > updateTimeMs ?
+            previousHostLagInfo : hostLagInfo);
+  }
+
+  /**
+   * Returns lag information for all of the "alive" hosts for a given state store and partition.
+   * @param lagInfoKey The lag info key
+   * @param partition The partition of that state
+   * @return A map which is keyed by host and contains lag information
+   */
+  public Map<HostInfo, LagInfoEntity> getHostsPartitionLagInfo(
+      final Set<HostInfo> hosts, final LagInfoKey lagInfoKey, final int partition) {
+    ImmutableMap.Builder<HostInfo, LagInfoEntity> builder = ImmutableMap.builder();
+    for (HostInfo host : hosts) {
+      LagInfoEntity lagInfo = receivedLagInfo.getOrDefault(host, EMPTY_HOST_LAG_INFO).getLagInfo()
+          .getOrDefault(lagInfoKey, Collections.emptyMap())
+          .getOrDefault(partition, null);
+      if (aliveHosts.containsKey(host) && lagInfo != null) {
+        builder.put(host, lagInfo);
+      }
+    }
+    return builder.build();
+  }
+
+  /**
+   * Returns a map of storeName -> partition -> LagInfoEntity.  Meant for being exposed in testing
+   * and debug resources.
+   */
+  public Map<String, Map<String, Map<Integer, LagInfoEntity>>> listAllLags() {
+      return receivedLagInfo.entrySet().stream()
+          .map(hostEntry -> Pair.of(
+              hostEntry.getKey().host() + ":" + hostEntry.getKey().port(),
+              hostEntry.getValue().getLagInfo().entrySet().stream()
+                  .map(storeNameEntry -> Pair.<String, Map<Integer, LagInfoEntity>>of(
+                      storeNameEntry.getKey().toString(),
+                      storeNameEntry.getValue().entrySet().stream()
+                          .map(partitionEntry -> Pair.of(
+                                partitionEntry.getKey(), partitionEntry.getValue()))
+                          .collect(ImmutableSortedMap.toImmutableSortedMap(
+                              Integer::compare, Pair::getLeft, Pair::getRight))))
+                  .collect(ImmutableSortedMap.toImmutableSortedMap(
+                      Comparator.comparing(String::toString), Pair::getLeft, Pair::getRight))))
+          .collect(ImmutableSortedMap.toImmutableSortedMap(
+              Comparator.comparing(String::toString), Pair::getLeft, Pair::getRight));
+  }
+
+  @Override
+  public void onHostStatusUpdated(final Map<String, HostStatusEntity> hostsStatusMap) {
+    aliveHosts.clear();
```

Comment:
do we need to synchronize between this update and the send lag iterating over the map to send lag to other nodes?

Reply:
Yes, good call.  There's technically a race as it is here.

I changed it to create an immutable set and then updates an AtomicReference.